File: tests/sdks/python/containers/lowkey_vault_container.py
from __future__ import annotations

import os
import time

import requests
from azure.identity import DefaultAzureCredential
from azure.keyvault.secrets import SecretClient
from envilder.infrastructure.azure.azure_key_vault_secret_provider import (
    AzureKeyVaultSecretProvider,
)
from testcontainers.core.container import DockerContainer

_IMAGE = "nagyesta/lowkey-vault:7.1.61"
_HTTPS_PORT = 8443
_HTTP_PORT = 8080


class LowkeyVaultContainer:
    def __init__(self) -> None:
        self._container: DockerContainer | None = None
        self._vault_url: str = ""
        self._token_url: str = ""
        self._prev_identity_endpoint: str | None = None
        self._prev_identity_header: str | None = None

    def start(self) -> "LowkeyVaultContainer":
        print("\n[LowkeyVault] Starting container...")

        try:
            self._container = (
                DockerContainer(_IMAGE)
                .with_exposed_ports(_HTTPS_PORT, _HTTP_PORT)
                .with_env(
                    "LOWKEY_ARGS",
                    "--server.port=8443" " --LOWKEY_VAULT_RELAXED_PORTS=true",
                )
            )
            self._container.start()

            host = self._container.get_container_host_ip()
            https_port = self._container.get_exposed_port(_HTTPS_PORT)
            http_port = self._container.get_exposed_port(_HTTP_PORT)

            self._vault_url = f"https://{host}:{https_port}"
            self._token_url = (
                f"http://{host}:{http_port}" "/metadata/identity/oauth2/token"
            )

            self._wait_until_ready()

            self._prev_identity_endpoint = os.environ.get("IDENTITY_ENDPOINT")
            self._prev_identity_header = os.environ.get("IDENTITY_HEADER")
            os.environ["IDENTITY_ENDPOINT"] = self._token_url
            os.environ["IDENTITY_HEADER"] = "dummy"
        except Exception:
            self.stop()
            raise

        print(f"[LowkeyVault] Ready at: {self._vault_url}")
        return self

    def stop(self) -> None:
        try:
            if self._container:
                print("[LowkeyVault] Stopping container...")
                self._container.stop()
                self._container = None
        finally:
            if self._prev_identity_endpoint is None:
                os.environ.pop("IDENTITY_ENDPOINT", None)
            else:
                os.environ["IDENTITY_ENDPOINT"] = self._prev_identity_endpoint

            if self._prev_identity_header is None:
                os.environ.pop("IDENTITY_HEADER", None)
            else:
                os.environ["IDENTITY_HEADER"] = self._prev_identity_header

    @property
    def vault_url(self) -> str:
        return self._vault_url

    def create_secret_client(self) -> SecretClient:
        return SecretClient(
            vault_url=self._vault_url,
            credential=DefaultAzureCredential(),
            connection_verify=False,  # test-only: Lowkey uses self-signed TLS
            verify_challenge_resource=False,
            api_version="7.6",  # pin to version supported by Lowkey Vault
        )

    def create_provider(self) -> AzureKeyVaultSecretProvider:
        return AzureKeyVaultSecretProvider(self.create_secret_client())

    def _wait_until_ready(
        self,
        max_retries: int = 30,
        delay: float = 1.0,
    ) -> None:
        url = f"{self._vault_url}/ping"
        for attempt in range(max_retries):
            try:
                response = requests.get(
                    url,
                    timeout=2,
                    verify=False,  # test-only: self-signed TLS
                )
                if response.status_code == 200:
                    return
            except requests.RequestException as e:
                if attempt == max_retries - 1:
                    raise TimeoutError(
                        "LowkeyVault did not become ready"
                        f" after {max_retries} attempts"
                    ) from e

            if attempt < max_retries - 1:
                time.sleep(delay)

        raise TimeoutError(
            "LowkeyVault did not become ready" f" after {max_retries} attempts"
        )

File: tests/sdks/nodejs/containers/lowkey-vault-container.ts
import crypto from 'node:crypto';
import https from 'node:https';
import { DefaultAzureCredential } from '@azure/identity';
import { SecretClient } from '@azure/keyvault-secrets';
import { GenericContainer, type StartedTestContainer } from 'testcontainers';
import { AzureKeyVaultSecretProvider } from '../../../../src/sdks/nodejs/src/infrastructure/azure/azure-key-vault-secret-provider.js';

const LOWKEY_VAULT_IMAGE = 'nagyesta/lowkey-vault:7.1.61';
const HTTPS_PORT = 8443;
const HTTP_PORT = 8080;
const MAX_RETRIES = 30;
const RETRY_DELAY_MS = 1000;

export class LowkeyVaultTestContainer {
  private container: StartedTestContainer | null = null;
  private vaultUrl = '';
  private prevIdentityEndpoint: string | undefined;
  private prevIdentityHeader: string | undefined;
  private prevTlsRejectUnauthorized: string | undefined;

  async start(): Promise<LowkeyVaultTestContainer> {
    console.log('\n[LowkeyVault] Starting container...');

    this.prevIdentityEndpoint = process.env.IDENTITY_ENDPOINT;
    this.prevIdentityHeader = process.env.IDENTITY_HEADER;
    this.prevTlsRejectUnauthorized = process.env.NODE_TLS_REJECT_UNAUTHORIZED;

    try {
      const suffix = crypto.randomUUID().slice(0, 8);
      this.container = await new GenericContainer(LOWKEY_VAULT_IMAGE)
        .withName(`lowkey-vault-${suffix}`)
        .withExposedPorts(HTTPS_PORT, HTTP_PORT)
        .withEnvironment({
          LOWKEY_ARGS: `--server.port=${HTTPS_PORT} --LOWKEY_VAULT_RELAXED_PORTS=true`,
        })
        .start();

      const host = this.container.getHost();
      const httpsPort = this.container.getMappedPort(HTTPS_PORT);
      const httpPort = this.container.getMappedPort(HTTP_PORT);

      this.vaultUrl = `https://${host}:${httpsPort}`;
      const tokenUrl = `http://${host}:${httpPort}/metadata/identity/oauth2/token`;

      await this.waitUntilReady();

      process.env.IDENTITY_ENDPOINT = tokenUrl;
      process.env.IDENTITY_HEADER = 'dummy';
      process.env.NODE_TLS_REJECT_UNAUTHORIZED = '0';
    } catch (error) {
      await this.stop();
      throw error;
    }

    console.log(`[LowkeyVault] Ready at: ${this.vaultUrl}`);
    return this;
  }

  async stop(): Promise<void> {
    try {
      if (this.container) {
        console.log('[LowkeyVault] Stopping container...');
        await this.container.stop();
        this.container = null;
      }
    } finally {
      restoreEnv('IDENTITY_ENDPOINT', this.prevIdentityEndpoint);
      restoreEnv('IDENTITY_HEADER', this.prevIdentityHeader);
      restoreEnv(
        'NODE_TLS_REJECT_UNAUTHORIZED',
        this.prevTlsRejectUnauthorized,
      );
    }
  }

  getVaultUrl(): string {
    return this.vaultUrl;
  }

  createSecretClient(): SecretClient {
    return new SecretClient(this.vaultUrl, new DefaultAzureCredential(), {
      disableChallengeResourceVerification: true,
    });
  }

  createProvider(): AzureKeyVaultSecretProvider {
    return new AzureKeyVaultSecretProvider(this.createSecretClient());
  }

  private async waitUntilReady(): Promise<void> {
    const url = `${this.vaultUrl}/ping`;
    const agent = new https.Agent({ rejectUnauthorized: false });

    for (let attempt = 0; attempt < MAX_RETRIES; attempt++) {
      try {
        const ok = await new Promise<boolean>((resolve) => {
          const req = https.get(url, { agent }, (res) => {
            resolve(res.statusCode === 200);
          });
          req.on('error', () => resolve(false));
          req.setTimeout(2000, () => {
            req.destroy();
            resolve(false);
          });
        });
        if (ok) {
          return;
        }
      } catch {
        // retry
      }

      if (attempt < MAX_RETRIES - 1) {
        await new Promise((r) => setTimeout(r, RETRY_DELAY_MS));
      }
    }

    throw new Error(
      `LowkeyVault did not become ready after ${MAX_RETRIES} attempts`,
    );
  }
}

function restoreEnv(name: string, previous: string | undefined): void {
  if (previous === undefined) {
    delete process.env[name];
  } else {
    process.env[name] = previous;
  }
}
